arangod/Aql/QueryList.cpp

```diff
@@ -262,85 +265,93 @@ void QueryList::remove(Query& query) {
 
 /// @brief kills a query
 Result QueryList::kill(TRI_voc_tick_t id) {
-  size_t const maxLength =
-      _maxQueryStringLength.load(std::memory_order_relaxed);
-
-  READ_LOCKER(writeLocker, _lock);
-
-  auto it = _current.find(id);
-
-  if (it == _current.end()) {
-    return {TRI_ERROR_QUERY_NOT_FOUND, "query ID not found in query list"};
+  // We should call dtor std::shared_ptr<Query> not under a lock,
+  // otherwise a deadlock will occur. Query dtor make a WRITE_LOCKER(..., _lock)
+  std::shared_ptr<Query> queryPtr;
+  {
+    READ_LOCKER(writeLocker, _lock);
+    auto it = _current.find(id);
+    if (it == _current.end()) {
+      return {TRI_ERROR_QUERY_NOT_FOUND, "query ID not found in query list"};
+    }
+    queryPtr = it->second.lock();
   }
-
-  killQuery(*it->second, maxLength, false);
-
-  return Result();
+  if (queryPtr) {
+    auto const length = _maxQueryStringLength.load(std::memory_order_relaxed);
+    killQuery(*queryPtr, length, false);
+  }
```

Comment:
Dunno I just keep the old behavior

Reply:
Yes, I think you're right

Reply:
Maybe not.
Now `killQuery` just write some log and call doing nothing method if query already killed

Reply:
True, but strictly speaking this is a change in behavior, since previously this situation simply could not occur. I think both versions are fine, I just think that it might be more consistent to also trigger the error in case we cannot lock the weak_ptr.